build-tools/xaprepare/xaprepare/Steps/Step_GenerateCGManifest.cs

```diff
@@ -0,0 +1,302 @@
+		public  abstract    void    WriteJsonTo (StreamWriter writer, HashSet<string> developmentDependencies);
+
+		protected CGManifestEntry ()
+		{
+		}
+	}
+
+	sealed class GitSubmoduleInfo : CGManifestEntry {
+
+		public  override    string      Name {
+			get {
+				const string github = "github.com/";
+				int i = RepositoryUrl.IndexOf (github, StringComparison.OrdinalIgnoreCase);
+				if (i >= 0)
+					return RepositoryUrl.Substring (i + github.Length);
+				return RepositoryUrl;
+			}
+		}
+
+		public  string      RepositoryUrl   {get; private set;}
+		public  string      CommitHash      {get; private set;}
+
+		private GitSubmoduleInfo ()
+		{
+		}
+
+		public override void WriteJsonTo (StreamWriter writer, HashSet<string> developmentDependencies)
+		{
+			string dev = developmentDependencies.Contains (Name)
+				? "true"
+				: "false";
+
+			writer.WriteLine ($"        {{");
+			writer.WriteLine ($"            \"Component\": {{");
+			writer.WriteLine ($"                \"Type\": \"git\",");
+			writer.WriteLine ($"                \"Git\": {{");
+			writer.WriteLine ($"                    \"RepositoryUrl\": \"{RepositoryUrl}\",");
+			writer.WriteLine ($"                    \"CommitHash\": \"{CommitHash}\"");
+			writer.WriteLine ($"                }}");
+			writer.WriteLine ($"            }},");
+			writer.WriteLine ($"            \"DevelopmentDependency\":{dev}");
+			writer.Write     ($"        }}");
+		}
+
+		const string Submodule = "submodule.external/";
+
+		public static IEnumerable<GitSubmoduleInfo> GetGitSubmodules (List<string> config, List<string> submoduleStatus)
+		{
+			string? entryId = null;
+			string? path = null;
+			string? url = null;
+
+			foreach (var line in config) {
+				if (!line.StartsWith (Submodule, StringComparison.Ordinal))
+					continue;
+
+				string? id = GetSubmoduleId (line);
+				if (id != entryId) {
+					if (url != null && path != null)
+						yield return CreateSubmoduleInfo (url, path);
+
+					entryId = id;
+					path    = null;
+					url     = null;
+				}
+
+				const string Path   = ".path=";
+				const string Url    = ".url=";
+				const string Git    = ".git";
+
+				int pathIndex   = line.IndexOf (Path, StringComparison.Ordinal);
+				if (pathIndex > 0) {
+					path        = line.Substring (pathIndex + Path.Length);
+					continue;
+				}
+
+				int urlIndex    = line.IndexOf (Url, StringComparison.Ordinal);
+				if (urlIndex > 0) {
+					int start   = urlIndex + Url.Length;
+					int count   = line.Length - start;
+					if (line.EndsWith (Git, StringComparison.Ordinal))
+						count -= Git.Length;
+					url         = line.Substring (start, count);
+					continue;
+				}
+			}
+
+			if (url != null && path != null)
+				yield return CreateSubmoduleInfo (url, path);
+
+			GitSubmoduleInfo CreateSubmoduleInfo (string url, string path)
+			{
+				string? hash = null;
+				foreach (var e in submoduleStatus) {
+					int pi  = e.IndexOf (path, StringComparison.Ordinal);
```

Comment:
I think `StringComparison.OrdinalIgnoreCase` would be safer here - URLs are not case-sensitive and the status paths come from the filesystem which may have a different case (especially on case-aware but case-insensitive filesystems)

Reply:
What this assumes is *consistency* between `git config --blob HEAD:.gitmodules --list` and `git submodule status`:

```bash
$ git config --blob HEAD:.gitmodules --list
…
submodule.external/Java.Interop.path=external/Java.Interop
submodule.external/Java.Interop.url=https://github.com/xamarin/java.interop.git
submodule.external/Java.Interop.branch=master
…

$ git submodule status
 60efc990cf7d33a55ff7864f215569d14ce51e34 external/Java.Interop (remotes/origin/HEAD-4-g60efc99)
…
```

Here, `path` would be `external/Java.Interop`, which is then "keyed" to the appropriate `git submodule status` line.

I *hope*/*assume* that these should be identical between the two commands, in which case `.Ordinal` should be sufficient.

The `url` isn't relevant here, and is "passed through as-is".

Reply:
`.gitmodules` is just a text file, it can potentially have files using any case, especially if created on a case-insensitive host. `git submodule status` reads the file, but it also reads the filesystem (reading the `.git` dir for information about the submodules) and who knows which case it will use if there's a discrepancy. It's just a precaution that doesn't cost much but may save a minor headache.

Reply:
Done; it uses `.OrdinalIgnoreCase` now.